src/librustdoc/core.rs

```diff
@@ -174,10 +177,30 @@ pub(crate) fn new_dcx(
     rustc_errors::DiagCtxt::new(emitter).with_flags(unstable_opts.dcx_flags(true))
 }
 
+/// Create the input (string or file path)
+pub(crate) fn make_input(
+    early_dcx: &EarlyDiagCtxt,
+    input: &String,
+) -> Result<Input, ErrorGuaranteed> {
+    Ok(if input == "-" {
+        let mut src = String::new();
+        if io::stdin().read_to_string(&mut src).is_err() {
+            // Immediately stop compilation if there was an issue reading
+            // the input (for example if the input stream is not UTF-8).
+            let reported =
+                early_dcx.early_err("couldn't read from stdin, as it did not contain valid UTF-8");
+            return Err(reported);
+        }
+        Input::Str { name: FileName::anon_source_code(&src), input: src }
+    } else {
+        Input::File(PathBuf::from(input))
+    })
+}
+
 /// Parse, resolve, and typecheck the given crate.
 pub(crate) fn create_config(
     RustdocOptions {
-        input,
+        input: _,
```

Comment:
Have you checked if it's possible to restructure the code in such a way that we can have `RustdocOptions.input : Input`? Might make the code “nicer”, not sure.

Reply:
I think I had actually tried that at first but quickly decided against, but I don't remember why.

Anyway, I tried it again, a apart from needing to add `#[derive(Clone)]` to `Input`, it when relatively smoothly.
It's quite "nice" now :smile: I like it better.